### PR TITLE
Add a collections package for using with streams / lambdas.

### PR DESCRIPTION
**Pull request by DenverM80**
Implement SourceConverter to move all logic out of the Source model constructor.  Pull in a few related functions from StructHelper.

Move ordering / filtering from CCodeGenerator into SourceConverter.

Update all tests for tweaked Source creation.  Also make some tests more targeted in testing output rather than end-to-end ApiContract parsing through code gen.

Add a RequestPrototype template, and process in ds3_h.ftl

Remove useless .keep files

### CODE REVIEW

File: ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/converters/SourceConverter.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
+package com.spectralogic.ds3autogen.c.converters;
```
Comment (RachelTucker) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/converters/SourceConverter.java]: missing copyright

File: ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/converters/SourceConverter.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import com.spectralogic.ds3autogen.c.models.*;
+import com.spectralogic.ds3autogen.c.models.Enum;
+import com.spectralogic.ds3autogen.utils.collections.GuavaCollectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.text.ParseException;
+
+public final class SourceConverter {
+    private static final Logger LOG = LoggerFactory.getLogger(RequestConverter.class);
```
Comment (RachelTucker) [inline ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/converters/SourceConverter.java]: Should be SourceConverter.class not RequestConverter.class

File: ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/SourceConverter_Test.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
+package com.spectralogic.ds3autogen.c;
```
Comment (RachelTucker) [inline ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/SourceConverter_Test.java]: missing copyright

File: ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/StructHelper_Test.java [hunk 4]
```diff
@@ -42,9 +43,8 @@
 import java.util.HashSet;
 import java.util.Set;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertTrue;
 import static junit.framework.TestCase.assertFalse;
+import static org.junit.Assert.assertTrue;
 
 public class StructHelper_Test {
     final static Logger LOG = LoggerFactory.getLogger(CCodeGenerator_Test.class);
```
Comment (RachelTucker) [inline ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/StructHelper_Test.java]: Should be StructHelper_Test.class

File: ds3-autogen-utils/src/main/java/com/spectralogic/ds3autogen/utils/collections/GuavaCollectors.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
+package com.spectralogic.ds3autogen.utils.collections;
```
Comment (RachelTucker) [inline ds3-autogen-utils/src/main/java/com/spectralogic/ds3autogen/utils/collections/GuavaCollectors.java]: missing copyright

File: ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java [hunk 6]
```diff
@@ -36,119 +37,133 @@
...
-        final Ds3SpecParser parser = new Ds3SpecParserImpl();
-        final Ds3ApiSpec spec = parser.getSpec(CCodeGenerator_Test.class.getResourceAsStream(inputSpecFile));
-
-        final ImmutableList<Request> allRequests = CCodeGenerator.getAllRequests(spec);
-        final ImmutableList<Enum> allEnums = CCodeGenerator.getAllEnums(spec);
-        final ImmutableSet<String> enumNames = EnumHelper.getEnumNamesSet(allEnums);
-        final ImmutableList<Struct> allStructs = CCodeGenerator.getAllStructs(spec, enumNames, allRequests);
-        final ImmutableList<Struct> allOrderedStructs = StructHelper.getStructsOrderedList(allStructs, enumNames);
-        final Source source = new Source(allEnums, allOrderedStructs, CCodeGenerator.getAllRequests(spec));
+        Map<String,Object> testMap = new HashMap<>();
```
Comment (rpmoore) [inline ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java]: Missing final

File: ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java [hunk 7]
```diff
@@ -36,119 +37,133 @@
...
 
     @Test
     public void testGenerateSimpleTypedefStruct() throws IOException, ParserException, TypeRenamingConflictException, ResponseTypeNotFoundException, ParseException, TemplateModelException {
-        final String inputSpecFile = "/input/SimpleTypedefStruct.xml";
         final TestFileUtilsImpl fileUtils = new TestFileUtilsImpl();
-        final Ds3SpecParser parser = new Ds3SpecParserImpl();
-        final Ds3ApiSpec spec = parser.getSpec(CCodeGenerator_Test.class.getResourceAsStream(inputSpecFile));
-
-        final Header header = new Header(CCodeGenerator.getAllEnums(spec), CCodeGenerator.getAllStructs(spec, ImmutableSet.of(), ImmutableList.of()), CCodeGenerator.getAllRequests(spec));
+        Map<String,Object> testMap = new HashMap<>();
```
Comment (rpmoore) [inline ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java]: Missing final

File: ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGeneratorAmazonS3Requests_Test.java [hunk 8]
```diff
@@ -207,38 +217,33 @@ public void testGenerateAmazonS3GetBucketRequest() throws IOException, ParserExc
...
-    // ds3_error* ds3_get_object(const ds3_client* client, const ds3_request* request, void* user_data, size_t(*callback)(void*,size_t, size_t, void*)) {
-
     @Test
-    public void testGenerateAmazonS3GetObjectRequest() throws IOException, ParserException, ResponseTypeNotFoundException, TypeRenamingConflictException, ParseException {
-        final String inputSpecFile = "/input/AmazonS3GetObjectRequest.xml";
+    public void testGenerateAmazonS3GetBucketRequestPrototype() throws IOException, TemplateModelException {
         final TestFileUtilsImpl fileUtils = new TestFileUtilsImpl();
-        final Ds3SpecParser parser = new Ds3SpecParserImpl();
-        final Ds3ApiSpec spec = parser.getSpec(CCodeGenerator_Test.class.getResourceAsStream(inputSpecFile));
+        Map<String,Object> testMap = new HashMap<>();
```
Comment (rpmoore) [inline ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGeneratorAmazonS3Requests_Test.java]: Missing final

File: ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGeneratorSpectraS3Requests_Test.java [hunk 9]
```diff
@@ -134,4 +150,33 @@ public void testGenerateSpectraS3GetBucketsRequest() throws IOException, ParserE
         assertTrue(output.contains("    return _parse_ds3_bucket_list_response(client->log, response);"));
         assertTrue(output.contains("}"));
     }
+
+    @Test
+    public void testGenerateSpectraS3GetBucketRequestPrototype() throws IOException, TemplateModelException {
+        final TestFileUtilsImpl fileUtils = new TestFileUtilsImpl();
+        Map<String,Object> testMap = new HashMap<>();
```
Comment (rpmoore) [inline ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGeneratorSpectraS3Requests_Test.java]: Missing final

File: ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java [hunk 10]
```diff
@@ -36,119 +37,133 @@
...
+        assertEquals(expectedOutput, output);
     }
 
     @Test
     public void testGenerateComplexTypedefStruct() throws IOException, ParserException, TypeRenamingConflictException, ResponseTypeNotFoundException, ParseException, TemplateModelException {
-        final String inputSpecFile = "/input/ComplexTypedefStruct.xml";
         final TestFileUtilsImpl fileUtils = new TestFileUtilsImpl();
-        final Ds3SpecParser parser = new Ds3SpecParserImpl();
-        final Ds3ApiSpec spec = parser.getSpec(CCodeGenerator_Test.class.getResourceAsStream(inputSpecFile));
+        Map<String,Object> testMap = new HashMap<>();
```
Comment (rpmoore) [inline ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java]: Missing final

**Comment by DenverM80** [comment]
updated

File: ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java [hunk 11]
```diff
@@ -36,119 +37,133 @@
...
-    public void testGenerateSingleTypedefEnum() throws IOException, ParserException, ResponseTypeNotFoundException, TypeRenamingConflictException, ParseException, TemplateModelException {
-        final String inputSpecFile = "/input/TypedefEnum.xml";
+    public void testGenerateSingleTypedefEnum() throws IOException, TemplateModelException {
         final TestFileUtilsImpl fileUtils = new TestFileUtilsImpl();
-        final Ds3SpecParser parser = new Ds3SpecParserImpl();
-        final Ds3ApiSpec spec = parser.getSpec(CCodeGenerator_Test.class.getResourceAsStream(inputSpecFile));
-
-        final Header header = new Header(CCodeGenerator.getAllEnums(spec), CCodeGenerator.getAllStructs(spec, ImmutableSet.of(), ImmutableList.of()), CCodeGenerator.getAllRequests(spec));
+        final Map<String,Object> testMap = new HashMap<>();
+        final Enum enumEntry = new Enum("ds3_job_status", ImmutableList.of("IN_PROGRESS", "COMPLETED", "CANCELED")); // "CANCELED" is intentionally spelled wrong, as in the API
```
Comment (rpmoore) [inline ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java]: The api misspells this?  Have you filed a bug for it?

**Comment by rpmoore** [comment]
I have one general comment.  What is the Source class?  It's not exactly clear what data it's holding.

**Comment by DenverM80** [comment]
I'll add a javadoc to the Source class.  Similar to the Header class, its modeling the ds3.c source file

**Comment by rpmoore** [comment]
I see.  Sounds good.  Once you do that and update the pull request due to it now being out of date I'll merge.

File: ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java (hunk 11 above)
Comment (DenverM80) [inline ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java]: I hadn't yet.  I thought there would be pushback to changing a value that's already "in the wild".  I'll file it now and others can decide if it deserves attention.

Comment (rpmoore) [inline ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java]: Thank you

Comment (DenverM80) [inline ds3-autogen-c/src/test/java/com/spectralogic/ds3autogen/c/CCodeGenerator_Test.java]: Turns out my editor thought I was british.  "Canceled" is the correct american spelling.
http://www.dictionary.com/browse/canceled?s=t

I'll remove the comment.

**Pull request merged**